src/ds++/VectorView.hh

```diff
@@ -0,0 +1,138 @@
+//----------------------------------*-C++-*----------------------------------//
+/*!
+ * \file   ds++/VectorView.hh
+ * \author Kelly G. Thompson <kgt@lanl.gov>, Steve Nolen <sdnolen@lanl.gov>
+ * \date   Friday, Dec 13, 2019, 17:49 pm
+ * \brief  A view class that presents contiguous 1-D as multi-D.
+ * \note   Copyright (C) 2019 Triad National Security, LLC.
+ *         All rights reserved. */
+//---------------------------------------------------------------------------//
+
+#ifndef rtt_dsxx_VectorView_hh
+#define rtt_dsxx_VectorView_hh
+
+#include "Assert.hh"
+#include <array>
+#include <vector>
+
+namespace rtt_dsxx {
+
+//============================================================================//
+/*!
+ * \class VectorView
+ * \brief A wrapper for std::vector<T> that presents a compatible multi-D
+ *        'view'.
+ *
+ * \tparam T The data storage for the underlying std::vector<T>.
+ * \tparam Extents A variatic template list of integral values that represent
+ *         the max value of each dimension of the 'view'.
+ * \todo Add operator= to allow assignment operations.
+ *
+ * Example:
+ *
+ * \code
+ * vector<double> v1 = { 1,2,3,4,5,6,7,8,9 };
+ * VectorView<double,3,3,3> myvecview(v1);
+ * std::cout << myvecview(2,0,1) << endl;
+ *
+ * \endcode
+ */
+//============================================================================//
+
+template <typename T, unsigned... Extents> class VectorView {
+public:
+  //--------------------------------------------------------------------------//
+  /*! \brief Compute the 1-D index associated with the actual data storage from
+   *         provided multi-D index values.
+   *
+   * \tparam Indices variatic template that represents a set of integral values,
+   *         one for each dimension of the 'view'.
+   * \param indices a list of integral types that indicate the multi-D offset.
+   * \return An size_t index into the underlying std::vector<T>.
+   *
+   * \todo Once we adopt c++17, this logic can be simplified using 'fold
+   *       expressions'.
+   */
+  template <typename... Indices> size_t idx(Indices... indices) {
+    Insist(sizeof...(indices) == dim,
```

Comment:
Can this `Insist` be demoted to a DbC call that is only on in Debug mode? Less safe of course, but I worry about the performance cost of an `if` statement for every array access if `VectorView` is used deep in compute loops.